File: packages/gbnf-experiment/data/20260909T184722Z_089a4294/ported_implementation/gbnf/rules_builder/parse_space.py
from __future__ import annotations


def char_at(src: str, pos: int) -> str:
    """Mirrors JavaScript's out-of-bounds string indexing, which yields a falsy value."""
    return src[pos] if 0 <= pos < len(src) else ""


def parse_space(src: str, pos: int, newline_ok: bool) -> int:
    while char_at(src, pos) in (" ", "\t", "#") or (
        newline_ok and char_at(src, pos) in ("\r", "\n")
    ):
        if char_at(src, pos) == "#":
            while char_at(src, pos) and char_at(src, pos) not in ("\r", "\n"):
                pos += 1
        else:
            pos += 1
    return pos

File: proposed-projects/round-trip-experiment/reverse/20260909T031655Z_89df6247/20260910T063135Z_0babbdab/ported_implementation/src/rulesBuilder/parseSpace.ts
const at = (src: string, pos: number): string =>
  pos >= 0 && pos < src.length ? src[pos] : '';

export const parseSpace = (
  src: string,
  pos: number,
  newlineOk: boolean
): number => {
  while (
    [' ', '\t', '#'].includes(at(src, pos)) ||
    (newlineOk && ['\r', '\n'].includes(at(src, pos)))
  ) {
    if (at(src, pos) === '#') {
      while (at(src, pos) && !['\r', '\n'].includes(at(src, pos))) {
        pos += 1;
      }
    } else {
      pos += 1;
    }
  }
  return pos;
};
